test/position_embedding/position_embedding.cpp

```diff
@@ -0,0 +1,217 @@
+
+#ifndef TEST_ALIBI_VERBOSE
+#define TEST_ALIBI_VERBOSE 0
+#endif
+
+template <typename DataType>
+struct attention_score
+{
+    ck_tile::index_t rows, cols;
+    std::vector<DataType> pixels;
+
+    attention_score(ck_tile::index_t rows_,
+                    ck_tile::index_t cols_,
+                    DataType init_v_ = static_cast<DataType>(0))
+        : rows(rows_), cols(cols_), pixels(rows_ * cols_, init_v_)
+    {
+    }
+
+    auto& operator()(ck_tile::index_t i_row, ck_tile::index_t i_col)
+    {
+        return pixels[i_row * cols + i_col];
+    }
+
+    void print()
+    {
+        for(auto i_row = 0; i_row < rows; i_row++)
+        {
+            for(auto i_col = 0; i_col < cols; i_col++)
+            {
+                std::cout << pixels[i_row * cols + i_col] << " ";
+            }
+            std::cout << std::endl;
+        }
+    }
+};
+
+template <typename DataType, bool RowMajor = true>
+void alibi_traverse_with_slope(attention_score<DataType>& score,
+                               DataType slope,
+                               ck_tile::AlibiMode mode = ck_tile::AlibiMode::VERTICAL)
+{
+    using Alibi = ck_tile::Alibi<DataType, RowMajor>;
+    auto alibi  = Alibi{slope, score.rows, score.cols, mode};
+
+    for(ck_tile::index_t i_row = 0; i_row < score.rows; i_row++)
+    {
+        for(ck_tile::index_t i_col = 0; i_col < score.cols; i_col++)
+        {
+            alibi.update(score(i_row, i_col), i_row, i_col);
+        }
+    }
+}
+
+std::string alibi_mode_to_str(ck_tile::AlibiMode mode)
+{
+    if(mode == ck_tile::AlibiMode::VERTICAL)
+        return std::string("alibi_verti");
+    else if(mode == ck_tile::AlibiMode::FROM_TOP_LEFT)
+        return std::string("alibi_top-l");
+    else if(mode == ck_tile::AlibiMode::FROM_BOTTOM_RIGHT)
+        return std::string("alibi_bot-r");
+    return "";
+}
+
+template <typename DataType, bool RowMajor = true>
+bool test_alibi_traverse_with_slope(ck_tile::index_t rows,
+                                    ck_tile::index_t cols,
+                                    DataType slope,
+                                    ck_tile::AlibiMode mode,
+                                    const std::vector<DataType>& expected)
+{
+    attention_score<DataType> score{rows, cols};
+    alibi_traverse_with_slope<DataType, RowMajor>(score, slope, mode);
+
+    bool is_match = std::equal(score.pixels.begin(), score.pixels.end(), expected.begin());
+#if TEST_ALIBI_VERBOSE
+    std::cout << "---------" << alibi_mode_to_str(mode) << ", " << rows << "x" << cols << "("
+              << (RowMajor ? "row_major" : "col_major") << ")"
+              << (is_match ? ", valie:y" : ", valid:n") << std::endl;
+    score.print();
+#endif
+    return is_match;
+}
+
+template <typename DataType>
+bool test_alibi_slope_generation(ck_tile::index_t nheads, const std::vector<DataType>& expected)
+{
+    auto slopes = ck_tile::get_alibi_slopes<DataType>(nheads);
+
+    bool is_match = std::equal(slopes.begin(),
```

Comment:
do we have to check the size of `slopes` & `expected`?
```c++
bool is_match = (slopes.size() == expected.size())
                           && std::equal(...)
```

Reply:
let me refactor to use https://en.cppreference.com/w/cpp/algorithm/equal#Version_7 overload of `std::equal`